src/ast/semantic_analyser.cpp

```diff
@@ -1704,5 +1705,32 @@ void SemanticAnalyser::assign_map_type(const Map &map, const SizedType &type)
   }
 }
 
+bool SemanticAnalyser::is_context_access(Expression *expr, bool is_pointer_dereferenced)
+{
+  // Example:
+  //   1. args->filename;
+  //   2. ((struct perf_event_data*)ctx)->regs.ip;
+  //   3. $a = ((struct perf_event_data*)ctx); $a->sample_period;
+
+  if (auto *p = dynamic_cast<Builtin*>(expr))
+    return p->ident == "ctx" || p->ident == "args";
+  else if (auto *p = dynamic_cast<Variable*>(expr))
+    return is_context_access(variable_[p->ident], is_pointer_dereferenced);
+  else if (auto *p = dynamic_cast<Binop*>(expr))
+    return is_context_access(p->left, is_pointer_dereferenced) ||
+             is_context_access(p->right, is_pointer_dereferenced);
+  else if (auto *p = dynamic_cast<FieldAccess*>(expr))
+    return is_context_access(p->expr, is_pointer_dereferenced);
+  else if (auto *p = dynamic_cast<Cast*>(expr))
+    return is_context_access(p->expr, is_pointer_dereferenced);
+  else if (auto *p = dynamic_cast<Unop*>(expr))
+  {
+    // Exclude an access such as args->a->b
+    if ((p->op != Parser::token::MUL) || !is_pointer_dereferenced)
+      return is_context_access(p->expr, (p->op == Parser::token::MUL) | is_pointer_dereferenced);
```

Comment:
Should be `||` right?

Reply:
Yes, logical OR is appropriate. I'll fix it.